File: Packages/TriathlonEngine/Sources/TriathlonEngine/Predictions/CSS.swift
import Foundation

/// Critical Swim Speed (natation).
///
/// À partir de deux tests (typiquement 400 m et 200 m) :
/// `CSS (m/s) = (D_long − D_court) / (T_long − T_court)`
public enum CSS {

    /// Vitesse critique en m/s à partir de deux contre-la-montre.
    public static func speedMetersPerSec(
        longDistanceM: Double, longTimeSec: Double,
        shortDistanceM: Double, shortTimeSec: Double
    ) -> Double {
        precondition(longDistanceM > shortDistanceM && longTimeSec > shortTimeSec)
        return (longDistanceM - shortDistanceM) / (longTimeSec - shortTimeSec)
    }

    /// Allure critique en secondes / 100 m.
    public static func pacePer100m(
        longDistanceM: Double, longTimeSec: Double,
        shortDistanceM: Double, shortTimeSec: Double
    ) -> Double {
        let v = speedMetersPerSec(
            longDistanceM: longDistanceM, longTimeSec: longTimeSec,
            shortDistanceM: shortDistanceM, shortTimeSec: shortTimeSec
        )
        return 100.0 / v
    }

    /// Convertit une allure (s/100 m) en vitesse (m/s).
    public static func speed(fromPacePer100m pace: Double) -> Double { 100.0 / pace }

    /// Prédit un temps (s) sur une distance (m) à partir de la CSS,
    /// avec une légère dérive de fatigue au-delà de la distance critique.
    public static func predictTimeSeconds(cssPacePer100m pace: Double, distanceM d: Double) -> Double {
        let v = speed(fromPacePer100m: pace)
        // Facteur de fatigue : la vitesse tenue baisse ~2 %/doublement au-delà de 400 m.
        let ref = 400.0
        let drift = d > ref ? pow(d / ref, 0.02) : 1.0
        return d / v * drift
    }
}

File: web/src/engine/predictions/css.ts
// Portage de Predictions/CSS.swift — Critical Swim Speed.

export function speedMetersPerSec(
  longDistanceM: number,
  longTimeSec: number,
  shortDistanceM: number,
  shortTimeSec: number,
): number {
  if (!(longDistanceM > shortDistanceM && longTimeSec > shortTimeSec)) throw new Error('CSS: long > court');
  return (longDistanceM - shortDistanceM) / (longTimeSec - shortTimeSec);
}

/** Allure critique en secondes / 100 m. */
export function pacePer100m(
  longDistanceM: number,
  longTimeSec: number,
  shortDistanceM: number,
  shortTimeSec: number,
): number {
  const v = speedMetersPerSec(longDistanceM, longTimeSec, shortDistanceM, shortTimeSec);
  return 100.0 / v;
}

/** Convertit une allure (s/100 m) en vitesse (m/s). */
export function speedFromPacePer100m(pace: number): number {
  return 100.0 / pace;
}

/** Prédit un temps (s) sur une distance (m) à partir de la CSS, avec dérive de fatigue. */
export function predictTimeSeconds(cssPacePer100m: number, distanceM: number): number {
  const v = speedFromPacePer100m(cssPacePer100m);
  const ref = 400.0;
  const drift = distanceM > ref ? Math.pow(distanceM / ref, 0.02) : 1.0;
  return (distanceM / v) * drift;
}

export const CSS = { speedMetersPerSec, pacePer100m, speedFromPacePer100m, predictTimeSeconds };
